sdk/core/azure-core/inc/azure/core/operation.hpp

```diff
@@ -0,0 +1,61 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// SPDX-License-Identifier: MIT
+
+#pragma once
+
+#include "azure/core/context.hpp"
+#include "azure/core/operation_status.hpp"
+#include "azure/core/response.hpp"
+
+#include <chrono>
+#include <memory>
+#include <string>
+
+namespace Azure { namespace Core {
+
+  //@param <T> The type of the final result of the long-running operation.
+  template <class T> class Operation {
+  private:
+    virtual std::unique_ptr<Http::RawResponse> PollInternal(Context& context) = 0;
+    virtual Response<T> PollUntilDoneInternal(std::chrono::milliseconds period, Context& context)
+        = 0;
+
+  protected:
+    OperationStatus m_status = OperationStatus::NotStarted;
+
+  public:
+    OperationStatus Status() const { return m_status; }
+
+    bool IsDone() const
+    {
+      return (
+          m_status == OperationStatus::Succeeded || m_status == OperationStatus::Cancelled
+          || m_status == OperationStatus::Failed);
+    }
+
+    bool HasValue() const { return (m_status == OperationStatus::Succeeded); }
+
+    virtual T Value() const = 0;
+
+    virtual std::string GetResumeToken() const = 0;
+
+    std::unique_ptr<Http::RawResponse> Poll()
+    {
+      Context context;
+      return PollInternal(context);
+    }
+
+    std::unique_ptr<Http::RawResponse> Poll(Context& context) { return PollInternal(context); }
```

Comment:
I don't understand why we are accepting context if we are not going to do anything with it. We should consider checking or `context.IsCanceled()` and setting the operationstatus to cancelled before calling `PollInternal` and not call it for such cases.

**Is the point that context is just a pass through and it's the implementer's responsibility to do with it what they wish?** Is there any reason why they wouldn't honor `context.IsCanceled()`?

I don't know what we'd do for returning the raw response type in that case, but something like:
```C+
    std::unique_ptr<Http::RawResponse> Poll(Context& context)
    {
      if (conext.IsCanceled())
      {
        m_status = OperationStatus::Cancelled;
        return ???; // or do we still end u p calling the user function? both are odd
      }
      return PollInternal(context);
    }
```
Same approach for `PollUntilDone`.

Reply:
This is a Mocked up Operation<T> for test purposes.  In real implementations the PollInternal method would trigger the pipeline via the client, i.e. pipeline->Send(context, ...._) method.  The pipeline is the place we check the context and the pipeline throws if the context is cancelled.

What I've written for the test is to eliminate the client and pipeline from my test implementation of Operation.  They are extra dependencies that the test doesn't need to validate an Operation builds, runs, functions as expected

Reply:
The Poll function makes an I/O request to the service. ALL service I/O requests take a context and honor it already. So, the derive type that implements Poll will take the Context and pass it to the pipeline which will honor it; there is no need for anything in Operation to check the Context.

In short, the context should be checked immediately before an I/O operation and we have 2 kinds of I/O operations. Just before making an HTTP request and we're going to have the Transport policy check the context for this. The other is when reading from a BodyStream and we're going to have BodyStream's Read method do this check.

We should not be checking context cancellation anywhere else in the SDK. 
Someday, we might have 1 exception to this: we might check Cancellation in a long-running compute-bound loop that does no I/O operations. Today, we don't have any of these I don't think we ever will.

Reply:
> The pipeline is the place we check the context and the pipeline throws if the context is cancelled.

> The Poll function makes an I/O request to the service. ALL service I/O requests take a context and honor it already. So, the derive type that implements Poll will take the Context and pass it to the pipeline which will honor it; there is no need for anything in Operation to check the Context.

This is a really important design detail for folks like me to understand the relationship between context and operation. Thank you!

I am definitely interested in seeing how the concrete derived types end up looking like in the client library and how  this all flows through the SDK.